### DISCUSSION TITLE
Managing time-consuming actions during the initialization of Tauri Apps with Native DB

### DISCUSSION THREAD

**Discussion by vincent-herlemont** [tauri-apps/tauri#8489] [Q&A] [answered]
I have released an embedded database [`native_db`](https://github.com/vincent-herlemont/native_db) and I've had [requests](https://github.com/vincent-herlemont/native_db/discussions/69) to use it with Tauri, which I have no experience with. I made a small example with Tauri, and in just 15 minutes, I was able to create a Tauri app with persistence + Native DB, which is amazing with very little code! Tauri is really great! (source code [`native_db_tauri_vanilla`](https://github.com/vincent-herlemont/native_db_tauri_vanilla))

Now, let's get to the heart of the matter: Native DB has an [automatic migration system](https://docs.rs/native_db/latest/native_db/transaction/struct.RwTransaction.html#method.migrate) for models. It's best to do this at the initialization of the application, and it can take time, depending on the size of the user data. In my example, I naively put it in the app's `setup` here [native_db_tauri_vanilla/main.rs#L94-L101](https://github.com/vincent-herlemont/native_db_tauri_vanilla/blob/0c0e43a7b96ca54aef3a69e4006bff087a030f30/src-tauri/src/main.rs#L94-L101). But I'm almost sure there's a best practice. Could you give me some advice on this point?

**Comment by FabianLars** [comment, marked as answer]
> I put it in the app's `setup` [...] I'm sure there's a best practice

I'd call putting it into the setup hook best practice. If you don't need access to anything from the `App`/`Window` instances doing it before the tauri::Builder line is pretty much the same though.

If you want to migrate the db and start the app at the same time i'd spawn a thread in the setup hook and would likely move the db builder stuff into it too and also get an `AppHandle` via `app.handle()` to be able to set the state in a command, that way you can get rid of the Arc (since tauri::State is basically one already)

If you don't want it to block the loading of the app you could spawn a thread in the setup hook

**Reply by vincent-herlemont** [reply]
Thank you for this valuable information, I was able to [improve the example](https://github.com/vincent-herlemont/native_db_tauri_vanilla/commit/b9ecf3d7f086b87348c06708efbbbefefa80ead6), and it's becoming really elegant (in my opinion). I will let the users enhance it with threads, etc., according to their needs.
